perun-core/src/main/java/cz/metacentrum/perun/core/entry/ServicesManagerEntry.java

```diff
@@ -60,6 +63,227 @@ public ServicesManagerEntry(PerunBl perunBl) {
+				for (Facility facility : destinationFacilities) {
+					if (AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility)) {
+						isFacilityAdminAtLeastOnce = true;
+					}
+				}
+				if (!isFacilityAdminAtLeastOnce && !AuthzResolver.isAuthorized(sess, Role.PERUNADMIN)) {
+					log.warn("Trying to unblock services on destination by a user who is neither perunadmin nor admin of the facility");
+					continue; //skip this one because of not authorized action
+				}
+				getServicesManagerBl().unblockAllServicesOnDestination(sess, destination.getId());
+			}
+		}
+	}
+
+	@Override
+	public void unblockAllServicesOnDestination(PerunSession sess, int destinationId) throws PrivilegeException, FacilityNotExistsException, DestinationNotExistsException {
+		Destination destination = getDestinationById(sess, destinationId);
+		List<Facility> destinationFacilities = perunBl.getFacilitiesManager().getFacilitiesByDestination(sess, destination.getDestination());
+
+		boolean isFacilityAdminAtLeastOnce = false;
+		for (Facility facility : destinationFacilities) {
+			if (AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility)) {
+				isFacilityAdminAtLeastOnce = true;
+			}
+		}
+		if (!isFacilityAdminAtLeastOnce && !AuthzResolver.isAuthorized(sess, Role.PERUNADMIN)) {
+			throw new PrivilegeException(sess, "unblockAllServicesOnDestination");
+		}
+		getServicesManagerBl().unblockAllServicesOnDestination(sess, destinationId);
+	}
+
+	@Override
+	public void unblockServiceOnFacility(PerunSession sess, Service service, Facility facility) throws PrivilegeException {
+		// Authorization
+		if (!AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility)) {
+			throw new PrivilegeException(sess, "unblockServiceOnFacility");
+		}
+		getServicesManagerBl().unblockServiceOnFacility(sess, service, facility);
+	}
+
+	@Override
+	public void unblockServiceOnDestination(PerunSession sess, Service service, int destinationId) throws PrivilegeException, FacilityNotExistsException, DestinationNotExistsException {
+		Destination destination = getDestinationById(sess, destinationId);
+		List<Facility> destinationFacilities = perunBl.getFacilitiesManager().getFacilitiesByDestination(sess, destination.getDestination());
+
+		boolean isFacilityAdminAtLeastOnce = false;
+		for (Facility facility : destinationFacilities) {
+			if (AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility)) {
+				isFacilityAdminAtLeastOnce = true;
+			}
+		}
+		if (!isFacilityAdminAtLeastOnce && !AuthzResolver.isAuthorized(sess, Role.PERUNADMIN)) {
+			throw new PrivilegeException(sess, "unblockServiceOnDestination");
+		}
+		getServicesManagerBl().unblockServiceOnDestination(sess, service, destinationId);
+	}
+
+	@Override
+	public boolean forceServicePropagation(PerunSession sess, Facility facility, Service service) throws PrivilegeException {
+		// Authorization
+		if (!AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, facility) &&
```

Comment:
Please remove ``PERUNOBSERVER`` from all ``plan/forceServicePropagation()`` methods (it is state changing operation) and add ``ENGINE`` since it is sometimes called from the gen scripts by engine.

Reply:
This is not yet resolved.